# Assignment 2: Entity Resolution (Part 2)


## Objective

In Assignment 2 (Part 2), you will learn how to use Active Learning to address the entity resolution problem. After completing this assignment, you should be able to answer the following questions:

1. Why Active Learning?
2. How to implement uncertain sampling, a popular query strategy for Active Learning?
3. How to solve an ER problem using Active Learning?


## Active Learning

[Active learning](http://tiny.cc/al-wiki) is a certain type of ML algorithms that can train a high-quality ML model with small data-labeling cost. Its basic idea is quite easy to understand. Consider a typical supervised ML problem, which requires a (relatively) large training dataset. In the training dataset, there may be only a small number of data points that are beneficial to the trained ML model. In other words, labeling a small number of data points is enough to train a high-quality ML model. The goal of active learning is to help us to identify those data points. 


In this assignment, we will develop an Active Learning approach for Entity Resolution. The following figure shows the architecture of an entity resolution solution. It consists of four major steps. **I will provide you the source code for Steps 1, 2, 4. Your job is to implement Step 3.**  

<img src="img/arch.png", width=800/>

![title](img/arch.png)

### Step 1. Read Data

Suppose we get a restaurant dataset `restaurant.csv`. The data has many duplicate restaurants.  For example, the first two rows shown below are duplicated (i.e., refer to the same real-world entity). You can check out all duplicate (matching) record pairs from `true_matches.json`. 

In [2]:
import pandas as pd

df = pd.read_csv('restaurant.csv')
data = df.values.tolist()
print("(#Rows, #Cols) :", df.shape)
df.head(5)

(#Rows, #Cols) : (858, 5)


id                       name  ...         city         type
0   1  arnie morton's of chicago  ...  los angeles     american
1   2  arnie morton's of chicago  ...  los angeles  steakhouses
2   3         art's delicatessen  ...  studio city     american
3   4                 art's deli  ...  studio city        delis
4   5              hotel bel-air  ...      bel air  californian

[5 rows x 5 columns]

### Step 2. Similar Pairs

We first use a similarity-join algorithm to generate similar pairs. 

Below is the code. After running the code, we get 678 similar pairs ordered by their similarity decreasingly.

In [3]:
from a2_utils import *

data = df.values.tolist()
simpairs = simjoin(data)

print("Num of Pairs: ", len(data)*(len(data)-1)/2)
print("Num of Similar Pairs: ", len(simpairs))
print("The Most Similar Pair: ", simpairs[0])

Num of Pairs:  367653.0
Num of Similar Pairs:  678
The Most Similar Pair:  ([170, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern/soul'], [169, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern'])


We can see that `simjoin` helps us remove the number of pairs from 367653 to 678. But, there are still many non-matching pairs in `simpairs` (see below). 

In [4]:
print(simpairs[-1])

([764, "buzio's in the rio", '3700 w. flamingo rd.', 'las vegas', 'seafood'], [542, 'carnival world', '3700 w. flamingo rd.', 'las vegas', 'buffets'])


Next, we will use active learning to train a classifier, and then use the classifier to classify each pair in `simpairs` as either "matching" or "nonmatching". 

### Step 3. Active Learning

Given a set of similar pairs, what you need to do next is to iteratively train a classifier to decide which pairs are truly matching. We are going to use [logistic regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) as our classifier. 

#### Initialization

At the beginning, all the pairs are unlabeled. To initialize a model, we first pick up ten pairs and then label each pair using  the `crowdsourcing()` function. You can assume that `crowdsourcing()` will ask a crowd worker (e.g., on Amazon Mechanical Turk) to label a pair. 


`crowdsourcing(pair)` is a function that simulates the use of crowdsourcing to label a pair
  
  - **Input:**	pair – A pair of records 

  - **Output:**	Boolean –  *True*: The pair of records are matching; *False*: The pair of records are NOT matching;

Please use the following code to do the initialization. 

In [5]:
from a2_utils import crowdsourcing

# choose the most/least similar five pairs as initial training data
init_pairs = simpairs[:5] + simpairs[-5:]
matches = []
nonmatches = []
for pair in init_pairs:
    is_match = crowdsourcing(pair)
    if is_match == True:
        matches.append(pair)
    else:
        nonmatches.append(pair)
        
print("Number of matches: ", len(matches))
print("Number of nonmatches: ", len(nonmatches))

Are they matching?
[170, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern/soul']
[169, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern']
Answer: Yes
Are they matching?
[88, 'manhattan ocean club', '57 w. 58th st.', 'new york city', 'seafood']
[87, 'manhattan ocean club', '57 w. 58th st.', 'new york', 'seafood']
Answer: Yes
Are they matching?
[112, 'san domenico', '240 central park s.', 'new york city', 'italian']
[111, 'san domenico', '240 central park s', 'new york', 'italian']
Answer: Yes
Are they matching?
[197, 'fleur de lys', '777 sutter st.', 'san francisco', 'french (new)']
[196, 'fleur de lys', '777 sutter st.', 'san francisco', 'french']
Answer: Yes
Are they matching?
[8, 'cafe bizou', '14016 ventura blvd.', 'sherman oaks', 'french bistro']
[7, 'cafe bizou', '14016 ventura blvd.', 'sherman oaks', 'french']
Answer: Yes
Are they matching?
[709, 'arcadia', '21 e. 62nd st.', 'new york city', 'american (new)']
[66, 'four seasons', '99 e. 5

Here is the only code you need to write in this assignment.


In [6]:
from a2_utils import featurize, crowdsourcing
from sklearn.linear_model import LogisticRegression

labeled_pairs = matches + nonmatches
unlabeled_pairs = [featurize(p) for p in simpairs if p not in labeled_pairs]
unlabel_unfea_pairs = [p for p in simpairs if p not in labeled_pairs]
iter_num = 5
df_predict=[]
labeled_pairs=[featurize(p) for p in labeled_pairs]
y=[1,1,1,1,1,0,0,0,0,0]
test_pairs=unlabeled_pairs

for iter in range(iter_num):
  df_predict=[]
  model=LogisticRegression().fit(labeled_pairs,y)
  df_predict=model.predict(test_pairs)
  df_prob=model.predict_proba(test_pairs)
  df_proba1=pd.DataFrame(df_prob)
  df_proba1.columns=['prob0', 'prob1']
  df_proba1["diff"]=abs(df_proba1["prob1"]-df_proba1["prob0"])
  df_proba1["pairs"]=[x for x in test_pairs]
  df_proba1["original_pairs"]=[x for x in unlabel_unfea_pairs]
  df_proba1=df_proba1.sort_values(ascending=False,by="diff")
  df_proba1=df_proba1.reset_index(drop=True)
  for i in range(10):
    is_match = crowdsourcing(df_proba1["original_pairs"][len(df_proba1)-i-1])
    if is_match == True:
        y.append(1)
        labeled_pairs.append(df_proba1["pairs"][len(test_pairs)-i-1])
        test_pairs.remove(df_proba1["pairs"][len(df_proba1)-i-1])
        unlabel_unfea_pairs.remove(df_proba1["original_pairs"][len(df_proba1)-i-1])
    else:
        y.append(0)
        labeled_pairs.append(df_proba1["pairs"][len(df_proba1)-i-1])
        test_pairs.remove(df_proba1["pairs"][len(df_proba1)-i-1])
        unlabel_unfea_pairs.remove(df_proba1["original_pairs"][len(df_proba1)-i-1])
  print(len(test_pairs))
  

#<-- Write Your Code -->

Are they matching?
[817, 'ritz-carlton cafe (atlanta)', '181 peachtree st.', 'atlanta', 'american (new)']
[180, 'ritz-carlton restaurant', '181 peachtree st.', 'atlanta', 'french (classic)']
Answer: No
Are they matching?
[555, "ralph's diner", '3000 las vegas blvd. s', 'las vegas', 'coffee shops/diners']
[538, "bugsy's diner", '3555 las vegas blvd. s', 'las vegas', 'coffee shops/diners']
Answer: No
Are they matching?
[4, "art's deli", '12224 ventura blvd.', 'studio city', 'delis']
[3, "art's delicatessen", '12224 ventura blvd.', 'studio city', 'american']
Answer: Yes
Are they matching?
[156, 'ciboulette restaurant', '1529 piedmont ave.', 'atlanta', 'french (new)']
[155, 'ciboulette', '1529 piedmont ave.', 'atlanta', 'french']
Answer: Yes
Are they matching?
[130, "virgil's real bbq", '152 w. 44th st.', 'new york city', 'bbq']
[129, "virgil's", '152 w. 44th st.', 'new york', 'american']
Answer: Yes
Are they matching?
[24, 'le chardonnay (los angeles)', '8284 melrose ave.', 'los angeles',

**[Algorithm Description].**   Active learning has many [query strategies](http://tiny.cc/al-wiki-qs) to decide which data point should be labeled. You need to implement uncertain sampling. The algorithm trains an initial model on `labeled_pairs`. Then, it iteratively trains a model. At each iteration, it first applies the model to `unlabeled_pairs`, and makes a prediction on each unlabeled pair along with a probability, where the probability indicates the confidence of the prediction. After that, it selects the most uncertain pair (If there is still a tie, break it randomly),  and call the `crowdsroucing()` function to label the pair. After the pair is labeled, it updates `labeled_pairs` and `unlabeled_pairs`, and then retrain the model on `labeled_pairs`.

**[Input].** 
- `labeled_pairs`: 10 labeled pairs (by default)
- `unlabeled_pairs`: 668 unlabeled pairs (by default)
- `iter_num`: 5 (by default)

**[Output].** 
- `model`: A logistic regression model built by scikit-learn


### Step 4. Model Evaluation

After training an model, you can use the following code to evalute it.

In [7]:
import json
from a2_utils import evaluate
import numpy as np
            
sp_features = np.array([featurize(sp) for sp in simpairs])
label = model.predict(sp_features)
pair_label = zip(simpairs, label)

identified_matches = []
for pair, label in pair_label:
    if label == 1:
        identified_matches.append(pair)
        
precision, recall, fscore = evaluate(identified_matches)

print("Precision:", precision)
print("Recall:", recall)
print("Fscore:", fscore)
   

Precision: 0.96
Recall: 0.9056603773584906
Fscore: 0.9320388349514563


## Submission

Complete the code in A2-2.ipynb, and submit it to the CourSys activity Assignment 2.